runner.go

```diff
@@ -470,15 +479,22 @@ func (runner *Runner) workerDone(info doneInfo) {
 		}
 		runner.params.Logger.Errorf("exited %q: %s", info.id, errStr)
 	}
-	if workerInfo.start == nil {
+	if workerInfo.start == nil && workerInfo.replaceStart == nil {
 		runner.params.Logger.Debugf("no restart, removing %q from known workers", info.id)
 
 		// The worker has been deliberately stopped;
 		// we can now remove it from the list of workers.
 		runner.removeWorker(info.id)
 		return
 	}
-	go runner.runWorker(workerInfo.restartDelay, info.id, workerInfo.start)
+	restartDelay := workerInfo.restartDelay
+	if workerInfo.replaceStart != nil {
+		runner.params.Logger.Debugf("worker %q is being replaced", info.id)
+		workerInfo.start = workerInfo.replaceStart
```

Comment:
this may need to be included in the mutex in the future if exported methods want to read .start or .replaceStart

Reply:
These are all only accessed by the single internal go routine that handles the processing of requests. Using the mutex in these internal functions would lead to a deadlock.